File: app/src/main/java/com/edison/librolink/ui/screen/home/HomeViewModel.kt
package com.edison.librolink.ui.screen.home

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.edison.librolink.data.retrofit.BookItem
import com.edison.librolink.data.retrofit.RetrofitInstance
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class HomeViewModel : ViewModel() {
    private val _books = MutableStateFlow<List<BookItem>>(emptyList())
    val books: StateFlow<List<BookItem>> = _books

    private var currentIndex = 0
    private val pageSize = 10 // Number of books to load per page
    private var isLoading = false

    fun loadMoreBooks(query: String) {
        if (isLoading) return

        isLoading = true
        viewModelScope.launch {
            try {
                val response = RetrofitInstance.api.searchBooks(query, startIndex = currentIndex)
                response.items?.let { newBooks ->
                    val uniqueBooks = newBooks.filter { newBook ->
                        _books.value.none { it.volumeInfo.title == newBook.volumeInfo.title }
                    }
                    _books.value += uniqueBooks // Append new books
                    currentIndex += pageSize
                }
            } catch (e: Exception) {
                Log.e("BooksViewModel", "Error loading books: ${e.message}")
            } finally {
                isLoading = false
            }
        }
    }

    fun clearBooks(){
        _books.value = emptyList()
        currentIndex = 0
    }
}

File: lib/ui/home/home_view_model.dart
import 'package:flutter/material.dart';
import 'package:dio/dio.dart';
import 'package:libro_link/data/retrofit/google_books_api.dart';
import 'package:libro_link/data/retrofit/book_response.dart';

class HomeViewModel extends ChangeNotifier {
  final GoogleBooksApi _api;
  final dio = Dio(BaseOptions(
    baseUrl: 'https://www.googleapis.com/books/v1/',
  ));
  HomeViewModel(Dio dio) : _api = GoogleBooksApi(dio);
  final List<BookItem> _books = [];

  List<BookItem> get books => _books;
  int _currentIndex = 0;
  final int _pageSize = 10;
  bool _isLoading = false;

  Future<void> loadMoreBooks(String query) async {
    if (_isLoading) return;
    _isLoading = true;
    try {
      final response = await _api.searchBooks(
        query: query,
        startIndex: _currentIndex,
        maxResults: _pageSize,
      );
      debugPrint(response.toString());
      final newBooks = response.items ?? [];
      final uniqueBooks = newBooks.where((newBook) =>
      !_books.any((existing) => existing.volumeInfo.title == newBook.volumeInfo.title)
      ).toList();
      _books.addAll(uniqueBooks);
      _currentIndex += _pageSize;
      notifyListeners();
    } catch (e) {
      debugPrint("Error loading books: $e");
    } finally {
      _isLoading = false;
    }
  }

  void clearBooks() {
    _books.clear();
    _currentIndex = 0;
    notifyListeners();
  }
}